### PR TITLE
Add checks for empty tensor list

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155383
* #155382

Vibe-coded with Codex, after collecting a backtrace, see https://chatgpt.com/s/cd_68438be8a1248191adbfa0a5f000e60b

Even though, check for empty tensor list exists in `at::cat` crash might happens while resolving named dimension to position, by calling `dimname_to_position(tensors[0], dim)`, see backtrace below
```
(lldb) up
frame #1: 0x00000001101146dc libtorch_cpu.dylib`at::TensorBase::has_names(this=0x0000000000000000) const at TensorBase.h:559:10
   556 	  bool has_names() const {
   557 	    // If a user is using unnamed tensors, then we can short-circuit right here.
   558 	    // Otherwise, impl::has_names attempts to retrieve names.
-> 559 	    if (!impl_->has_named_tensor_meta()) {
   560 	      return false;
   561 	    }
   562 	    return impl::has_names(unsafeGetTensorImpl());
(lldb) up
frame #2: 0x00000001101144c4 libtorch_cpu.dylib`at::dimname_to_position(tensor=0x0000000000000000, dim=Dimname @ 0x000000016fdfe348) at NamedTensorUtils.cpp:23:3
   20  	int64_t dimname_to_position(const Tensor& tensor, Dimname dim) {
   21  	  TORCH_CHECK(dim.type() != NameType::WILDCARD,
   22  	      "Please look up dimensions by name, got: name = None.");
-> 23  	  TORCH_CHECK(tensor.has_names(),
   24  	      "Name ", dim, " not found in ", toDimnameRepr(tensor), ".");
   25  	  const auto names = tensor.names();
   26  	
```

TODOs: 
 - May be move test from `test_tensor_creation.py` to OpInfo (not sure which one is more readable)
 - Replace  `TORCH_CHECK` with `TORCH_CHECK_VALUE` and adjust unit tests 
 
Fixes https://github.com/pytorch/pytorch/issues/155306

### CODE REVIEW

**Review by cyyever — APPROVED** [review]

**Comment by ezyang** [comment]
Can you confirm from the transcript that the tests fail before the change?

**Comment by ezyang** [comment]
Also the diff is kind of suspicious, all the places annotated with checks call functions I would also expect to do the non empty check

**Comment by malfet** [comment]
> Can you confirm from the transcript that the tests fail before the change?

Yes, run `python -c "import torch; torch.concat([], dim='N')"` now and it will crash. Add empty check to concat and it will throw the exception.

> Also the diff is kind of suspicious, all the places annotated with checks call functions I would also expect to do the non empty check

It crashes inside `dimname_to_position` that is called with `tensors[0]` as first argument(see backtrace below). If it's documented somewhere, that calling `tensor[0]` on empty tensor list guarantees to return nullptr, I can move the check there.

```
Process 88909 stopped
* thread #2, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000110114940 libtorch_cpu.dylib`c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>::operator->(this=0x0000000000000000) const at intrusive_ptr.h:423:12
   420 	  }
   421 	
   422 	  TTarget* operator->() const noexcept {
-> 423 	    return target_;
   424 	  }
   425 	
   426 	  operator bool() const noexcept {
Target 0: (Python) stopped.
(lldb) up
frame #1: 0x00000001101146dc libtorch_cpu.dylib`at::TensorBase::has_names(this=0x0000000000000000) const at TensorBase.h:559:10
   556 	  bool has_names() const {
   557 	    // If a user is using unnamed tensors, then we can short-circuit right here.
   558 	    // Otherwise, impl::has_names attempts to retrieve names.
-> 559 	    if (!impl_->has_named_tensor_meta()) {
   560 	      return false;
   561 	    }
   562 	    return impl::has_names(unsafeGetTensorImpl());
(lldb) up
frame #2: 0x00000001101144c4 libtorch_cpu.dylib`at::dimname_to_position(tensor=0x0000000000000000, dim=Dimname @ 0x000000016fdfe348) at NamedTensorUtils.cpp:23:3
   20  	int64_t dimname_to_position(const Tensor& tensor, Dimname dim) {
   21  	  TORCH_CHECK(dim.type() != NameType::WILDCARD,
   22  	      "Please look up dimensions by name, got: name = None.");
-> 23  	  TORCH_CHECK(tensor.has_names(),
   24  	      "Name ", dim, " not found in ", toDimnameRepr(tensor), ".");
   25  	  const auto names = tensor.names();
   26  	
(lldb) bt
* thread #2, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000014007f588 libtorch_cpu.dylib`at::dimname_to_position(at::Tensor const&, at::Dimname) + 40
    frame #1: 0x00000001407b87c0 libtorch_cpu.dylib`at::native::concat(c10::ArrayRef<at::Tensor>, at::Dimname) + 36
    frame #2: 0x00000001410f0c24 libtorch_cpu.dylib`at::_ops::concat_names::call(c10::ArrayRef<at::Tensor>, at::Dimname) + 468
    frame #3: 0x00000001028cef24 libtorch_python.dylib`torch::autograd::THPVariable_concat(_object*, _object*, _object*) + 944
```

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -774,10 +774,12 @@ Tensor cat(TensorList tensors, Dimname dim) {
 
 // torch.concat, alias for torch.cat
 Tensor& concat_out(TensorList tensors, Dimname dim, Tensor& result) {
+  TORCH_CHECK(!tensors.empty(), "expected a non-empty list of Tensors");
```
Comment (Skylion007) [inline aten/src/ATen/native/TensorShape.cpp:777]:
```suggestion
  TORCH_CHECK_VALUE(!tensors.empty(), "expected a non-empty list of Tensors");
```
These should all be ValueErrors

**Review by ezyang — APPROVED** [review]
I misunderstood the change, current location is good.

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (malfet) [reply aten/src/ATen/native/TensorShape.cpp:777]: I'll replace all of them in https://github.com/pytorch/pytorch/pull/155460
